staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go

```diff
@@ -253,3 +253,68 @@ func TestMicroTimeIsZero(t *testing.T) {
 		})
 	}
 }
+
+func TestMicroTimeUnmarshalJSONAndProtoEqual(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  MicroTime
+		result bool
+	}{
+		{"nanosecond level precision", UnixMicro(123, 123123123), true},
+		{"microsecond level precision", UnixMicro(123, 123123000), true},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			jsonData, err := c.input.MarshalJSON()
+			if err != nil {
+				t.Fatalf("Failed to marshal input to JSON: '%v': %v", c.input, err)
+			}
+
+			protoData, err := c.input.Marshal()
+			if err != nil {
+				t.Fatalf("Failed to marshal input to proto: '%v': %v", c.input, err)
+			}
+
+			var tJSON, tProto MicroTime
+			if err = tJSON.UnmarshalJSON(jsonData); err != nil {
+				t.Fatalf("Failed to unmarshal JSON: '%v': %v", jsonData, err)
+			}
+			if err = tProto.Unmarshal(protoData); err != nil {
+				t.Fatalf("Failed to unmarshal proto: '%v': %v", protoData, err)
+			}
+
+			result := tJSON.Equal(&tProto)
+			if result != c.result {
+				t.Errorf("Failed equality test for '%v': expected %+v, got %+v", c.input, c.result, result)
+			}
+		})
+	}
+}
+
+func TestMicroTimeProtoUnmarshalRaw(t *testing.T) {
+	cases := []struct {
+		name     string
+		input    []byte
+		expected MicroTime
+	}{
+		// input is generated by Timestamp{123, 123123123}.Marshal()
+		{"nanosecond level precision", []byte{8, 123, 16, 179, 235, 218, 58}, UnixMicro(123, 123123000)},
```

Comment:
lgtm, I verified that this unmarshaled to `UnixMicro(123, 123123123)` when the truncation in Unmarshal is commented out